Add specification of log correlation for tracing (issue #123).

sebright opened the pull request:
The specification only covers aspects of log correlation that are likely to be
shared by log correlation implementations for multiple languages and logging
frameworks.  It is based on the experimental log correlation libraries in
opencensus-java:

https://github.com/census-instrumentation/opencensus-java/tree/master/contrib/log_correlation/stackdriver
https://github.com/census-instrumentation/opencensus-java/tree/master/contrib/log_correlation/log4j2

/cc @ramonza @odeke-em @rghetia

sebright commented:
/cc @savaki

rakyll commented on trace/LogCorrelation.md:
```diff
@@ -0,0 +1,63 @@
+
+Log correlation is a feature that inserts information about the current span into log entries
+created by existing logging frameworks.  The feature can be used to add more context to log entries,
+filter log entries by trace ID, or show log entries as annotations on a trace.
```
"log entries as annotations on a trace"

This is a possibility but do we want to encourage this usage? I think we should keep it out.

adriancole replied:
I think the assumption that this would be used to decorate traces might be faulty. it would be more likely to have correlated queries especially as tracing uis are not as good at filtering logs (which can be extremely volumnous) than logging uis.

adriancole replied:
so I agree we should not anchor people with the idea that this will result in log entries as span annotations. eg leave it out or rename to language that is correlation in nature vs the word annotation which is historically the name of a part of a span

rakyll commented on trace/LogCorrelation.md:
```diff
@@ -0,0 +1,63 @@
+
+## String format for tracing data
+
+The logging framework may require the pieces of tracing data to be converted to strings.  In that
```
Tracing data or identifiers?

sebright replied:
Do you mean it should say "The logging framework may require the tracing data or identifiers to be converted to strings"?

rakyll commented on trace/LogCorrelation.md:
```diff
@@ -0,0 +1,63 @@
+
+Some logging frameworks allow the insertion of arbitrary key-value pairs into log entries.  When
+a log correlation implementation inserts tracing data by that method, the key names should be
+"opencensusTraceId", "opencensusSpanId", and "opencensusTraceSampled" by default.  The log
```
These are very mouthful key names.

OTOH, underscore is often preferred over camel case in log labels. We might want to reconsider these keys.

sebright replied:
> These are very mouthful key names.

There was some discussion about short vs long key names in https://github.com/census-instrumentation/opencensus-java/pull/1371#discussion_r211440499.

I thought that it would be better to use long key names by default, to avoid conflicting with other tracing libraries' key names.  Users could then override the key names if they wanted to purposefully use the same key names for, say, Brave and OpenCensus.

> OTOH, underscore is often preferred over camel case in log labels. We might want to reconsider these keys.

I'm happy to change the key names if underscores are more common.  Where have you seen underscores used?

I saw camel case logging context key names in Brave, Wingtips, and Spring Cloud Sleuth and camel case tracing-related log entry fields in Stackdriver (https://cloud.google.com/logging/docs/agent/configuration#special_fields_in_structured_payloads).  The three libraries are all Java, though, so it could easily be Java specific.

rakyll replied:
It could be a language-community preference. I saw it on Stackdriver Logging, zap and logrus.

https://github.com/uber-go/zap
https://github.com/sirupsen/logrus

ramonza replied:
In the unusual case of conflict, users can always override the defaults.

adriancole replied:
I think we should defer to language defaults. it would be weird to not use lower camel in most java logging formats (though not all). so we can make examples in a case format and caveat that appropriate use will occur when applied to a logging framework

bogdandrutu replied:
I think supporting per language name is not the best option:
1. Usually the log backends are not per language. So the backend has to understand all of these combinations.
2. I agree that size matters here so probably shorter is better (less data written).
3. I understand that for Java a key like "oc_trace_id" is probably not a good fit. For the moment we have only Java and Go here. Can we get some input from other languages?

Even if we have multiple keys we need to document them here for each language.

sebright replied:
I opened an issue for continuing the discussion: #195

adriancole commented on trace/LogCorrelation.md:
```diff
@@ -0,0 +1,70 @@
+## Tracing data to include in log entries
+
+A log correlation implementation should insert the following pieces of tracing data from the current
+span context into each log entry:
```
make available in the logging entry. eg in some systems it might not literally be a copy rather sharing in context.

adriancole commented on trace/LogCorrelation.md:
```diff
@@ -0,0 +1,70 @@
+
+The trace ID of the current span.  See [Span.md#traceid](Span.md#traceid).
+
+### Span ID
```
note it is common for some tooling to attempt to reverse engineer spans from log entries. when this is the case the parent id is also put in the logging context

bogdandrutu replied:
Maybe add a todo to consider this. Until we have a clear request it is not worth doing it.

sebright replied:
Done

adriancole commented on trace/LogCorrelation.md:
```diff
@@ -0,0 +1,70 @@
+"opencensusTraceId", "opencensusSpanId", and "opencensusTraceSampled" by default.  The log
+correlation implementation may allow the user to override the tracing data key names.
+
+## Deciding when to add tracing data to a log entry
```
I think this feature is a bit speculative. I would leave it out as I have only seen this practice here. usually it isnt selective.

SergeyKanzhelev commented on trace/LogCorrelation.md:
```diff
@@ -0,0 +1,70 @@
+The sampling bit of the current span, as a boolean.  See
+[Span.md#supported-bits](Span.md#supported-bits).
+
+## String format for tracing data
```
I suggest to add a section for "Other fields from tracestate" or something like this. So vendors-specific key/value pairs can be added to the logs.

bogdandrutu replied:
I think it worth mentioning the Tracestate here. Probably every integration should offer a callback mechanism to extract and log data from Tracestate.

sebright replied:
How about I add a TODO and open an issue for this part?  I think it will require more discussion, and this PR already has a lot of comments.

sebright replied:
I added a TODO for this feature.

sebright replied:
I opened #196.

bogdandrutu commented on trace/LogCorrelation.md:
```diff
@@ -0,0 +1,63 @@
+# Log Correlation
+
+Log correlation is a feature that inserts information about the current span into log entries
```
+1 for @ramonza.

bogdandrutu commented on trace/LogCorrelation.md:
```diff
@@ -0,0 +1,70 @@
+
+The span ID of the current span.  See [Span.md#spanid](Span.md#spanid).
+
+### Sampling Decision
```
@SergeyKanzhelev do you want to write a one pager explaining the samplingScore, how can be used, why is important to be propagated, etc.

For the moment I would not mention it here until we define it. We can leave a TODO to say when decide on sampling score then consider to add it here.

sebright replied:
I added a TODO.

sebright commented:
I reverted the changes to the key names for now, and I'm planning to open an issue for deciding on the key names separately.

sebright commented:
I labeled the specification as a draft and linked to the issue about deciding on key names.  I'll merge it now.

sebright commented:
I think I addressed all of the comments by updating the PR, adding TODOs, or opening issues, but please let me know if I missed anything.  Thanks for the reviews!
